strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/checksum/ChecksumCacheManager.java

```diff
@@ -1,226 +1,114 @@
 package org.carlspring.strongbox.storage.checksum;
 
-import java.util.LinkedHashMap;
-import java.util.LinkedHashSet;
+import org.carlspring.commons.encryption.EncryptionAlgorithmsEnum;
+import org.carlspring.strongbox.data.CacheName;
+import org.carlspring.strongbox.providers.io.RepositoryPath;
+
+import javax.inject.Inject;
+import java.util.Arrays;
+import java.util.List;
 import java.util.Map;
-import java.util.Optional;
-import java.util.Set;
+import java.util.Objects;
 
+import com.google.common.collect.ImmutableMap;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.DisposableBean;
+import org.springframework.cache.Cache;
+import org.springframework.cache.CacheManager;
+import org.springframework.stereotype.Component;
 
 /**
- * All artifacts should pass through here.
- * Any deployed file which doesn't end in a checksum format (md5, sha1, gpg)
- * should add a cachedChecksum. When an actual checksum is deployed,
- * (which is right after the artifact or metadata file has been deployed),
- * the cache should be queried for a match.
- * <p>
- * If:
- * - a match is found and matches, remove it from the cache. (If the checksums set
- * is empty, remove the respective Checksum from the cachedChecksums).
- * - a match is found, but does not match, trigger an event and log this, then remove
- * the checksum from the cache. (If the checksums set is empty, remove the respective
- * Checksum from the cachedChecksums).
- * - a checksum is not claimed within cachedChecksumLifetime, trigger an event and log
- * this, then remove the checksum from the cache. (If this checksums set is empty,
- * remove the respective Checksum from the cachedChecksums).
- *
- * @author mtodorov
+ * @author Przemysław Fusik
  */
+@Component
 public class ChecksumCacheManager
+        implements DisposableBean
```

Comment:
@fuss86 how you assume to invalidate this cache?

and also if you will invalidate it somehow then it will lose checksums that you need.

Reply:
There is no cache invalidation right now. We can think about it and prepare a follow-up task later for this , but right now it works this way: once the new cache value has been calculated - it replaces old one in cache.

Reply:
There won't be OOM because there is [default config setup for this cache](https://github.com/strongbox/strongbox/blob/issue-827/strongbox-data-service/src/main/java/org/carlspring/strongbox/config/HazelcastConfiguration.java#L45) so entries will be evicted when free heap size is below 1000 MB. Besides that checksums are cheap and cache is the fastest solution. 

However, I'm keen on re-implementing it now if you can provide some better alternative (having the code base we already have).

Reply:
I suggested the solution previously, you can directly store calculated checksums within artifact upload (instead of cache that you currenly use) and then overwrite it if original checksum uploaded.

Reply:
My initial implementation was such that when an artifact would get deployed:
* It's checksum would get added to the cache, based on the calculations of the `MultipleDigestInputStream`.
* Items in the cache would expire in a certain amount of time.
* The cache, (albeit quite primitive in my implementation), would remove entries, if they weren't accessed after a certain time (if I recall correctly).
* When an artifact's checksums were uploaded, they would get compared to the ones in the cache and:
  * If they matched, the checksum would get removed from the cache
  * If they didn't match, a warning would be printed out and _**then**_ checksum would be removed from the cache.

Obviously, this was a very primitive implementation, as it neither used EhCache, nor Hazelcast. And during the various refactorings this behavior may have changed.

@fuss86 : How does it work now?

Reply:
@carlspring I wrote about it above:

```
* Old ChecksumCacheManager has been renamed to CalculatedChecksumCacheManager (feel free to propose better name)
* I've introduced my own ChecksumCacheManager which I benefit from
* I haven't done anything else in the previous ChecksumCacheManager other than renaiming it (the same goes to the test class)
```

I see that it's all not clear so I'll change my implementation.

Reply:
> My initial implementation was such that when an artifact would get deployed:

I'll revert it all back.

Reply:
> I suggested the solution previously, you can directly store calculated checksums within artifact upload (instead of cache that you currenly use) and then overwrite it if original checksum uploaded.

OK

Reply:
> Or you can even just drop the cache, because your logic seems implies case when there is no checksum (which is the same as evicted from cache checksum).

"My" `ChecksumCacheManager` was used to handle expiration of the `RepositoryPath` so thiese checksums were used.

Reply:
> "My" ChecksumCacheManager was used to handle expiration of the RepositoryPath so these cached checksums were used.

sorry, but I dont understand what you mean here